src/intersection-observer-polyfill.js

```diff
@@ -0,0 +1,219 @@
+
+    /** @private {?Element} */
+    this.element_ = null;
+
+    /** @private @const {number} */
+    this.prevTime_ = -1;
+
+    /**
+     * The prev threshold slot which the previous ratio fills
+     * Range [0, this.threshold_.length]
+     * TODO: Do we always want to call callback at the first tick?
+     * @private {number}
+     */
+    this.prevThresholdSlot_ = 0;
+
+    // TODO: Add the PositionObserver here.
+
+    // TODO: Add the postMessageApi logic outside IntersectionObserverPolyfill.
+  }
+
+  /**
+   * Provide a way to observer the intersection change for a specific element
+   * @param {!Element} element
+   */
+  observe(element) {
+    // TODO: create the PositionObserver class, or create the listener for
+    // position info if inside an iframe.
+    this.element_ = element;
+  }
+
+  /**
+   * Tick function that PositionObserver can tell the class to calculate
+   * intersection.
+   * @param {!./layout-rect.LayoutRectDef} element element's rect
+   * @param {!./layout-rect.LayoutRectDef} viewport viewport's rect.
+   */
+  tick(element, viewport) {
+    const changeEntry =
+        this.getValidIntersectionChangeEntry_(element, viewport);
+    if (changeEntry) {
+      this.callback_(changeEntry);
+    }
+  }
+
+  /**
+   * Return a change entry for that should be compatible with
+   * IntersectionObserverEntry if it's valid with current config.
+   *
+   * Mutates passed in rootBounds to have x and y according to spec.
+   *
+   * @param {!./layout-rect.LayoutRectDef} element element's rect
+   * @param {!./layout-rect.LayoutRectDef} viewport viewport's rect.
+   * @return {?IntersectionObserverEntry} A valid change entry.
+   * @private
+   */
+  getValidIntersectionChangeEntry_(element, viewport) {
+    // Building an IntersectionObserverEntry.
+
+    let intersectionRect = element;
+    intersectionRect = rectIntersection(viewport, intersectionRect) ||
+        // No intersection.
+        layoutRectLtwh(0, 0, 0, 0);
+
+    // The element is relative to (0, 0), while the viewport moves. So, we must
+    // adjust.
+    // TODO(jridgewell, #5149): Fixed position elements must be recalculated.
+    const boundingClientRect = moveLayoutRect(element, -viewport.left,
+        -viewport.top);
+    intersectionRect = moveLayoutRect(intersectionRect, -viewport.left,
+        -viewport.top);
+
+    const ratio = intersectionRatio(intersectionRect, element);
+    const newThresholdSlot = this.getThresholdSlot_(ratio);
+    if (newThresholdSlot == this.prevThresholdSlot_) {
+      return null;
+    }
+    this.prevThresholdSlot_ = newThresholdSlot;
+
+    // Now, move the viewport to (0, 0)
+    const rootBounds = moveLayoutRect(viewport, -viewport.left, -viewport.top);
+
+    if (Date.now() == this.prevTime_) {
+      // TODO: Is this really needed?
+      return null;
+    }
+    this.prevTime_ = Date.now();
+    return /** @type {!IntersectionObserverEntry} */ ({
+      time: Date.now(),
```

Comment:
changed. However I see we don't use `performance.now()` anywhere. It seems to be a widely supported function. It's only because we never get into the case that we need time data from load?